ember-server/src/main/scala/org/http4s/ember/server/internal/ServerHelpers.scala

```diff
@@ -175,40 +182,82 @@ private[server] object ServerHelpers {
       onWriteFailure: (Option[Request[F]], Response[F], Throwable) => F[Unit]
   ): Stream[F, Nothing] = {
     val _ = logger
-    Stream
-      .unfoldLoopEval(socket.reads(receiveBufferSize, durationToFinite(idleTimeout)))(s =>
-        runApp(s, maxHeaderSize, requestHeaderReceiveTimeout, httpApp, errorHandler).attempt.map {
-          case Right((req, resp, rest)) => (Right((req, resp)), Some(rest))
-          case Left(e) => (Left(e), None)
-        })
-      .evalMap {
-        case Right((req, resp)) =>
-          postProcessResponse(req, resp).map(resp => (req, resp).asRight[Throwable])
-        case other => other.pure[F]
-      }
-      .evalTap {
-        case Right((request, response)) =>
-          send(socket)(Some(request), response, idleTimeout, onWriteFailure)
-        case Left(err) =>
-          err match {
-            case req: Parser.Request.ReqPrelude.ParsePreludeError
-                if req == Parser.Request.ReqPrelude.emptyStreamError =>
-              Applicative[F].unit
-            case err =>
-              errorHandler(err)
-                .handleError(_ => serverFailure.covary[F])
-                .flatMap(send(socket)(None, _, idleTimeout, onWriteFailure))
-          }
-      }
-      .takeWhile {
-        case Left(_) => false
-        case Right((req, resp)) =>
-          !(
-            req.headers.get(Connection).exists(_.hasClose) ||
-              resp.headers.get(Connection).exists(_.hasClose)
-          )
-      }
-      .drain ++ Stream.eval_(socket.close)
+
+    Stream.eval(mkRequestVault(socket)).flatMap { requestVault =>
+      Stream
+        .unfoldLoopEval(socket.reads(receiveBufferSize, durationToFinite(idleTimeout)))(s =>
+          runApp(
+            s,
+            maxHeaderSize,
+            requestHeaderReceiveTimeout,
+            httpApp,
+            errorHandler,
+            requestVault).attempt.map {
```

Comment:
Adding `requestVault` is the only change in this hunk.  The rest is indentation from the flatMapping of `requestVault`